### PR TITLE
unr-4845 Refactor out sender logic

### PR DESCRIPTION
**Pull request by jacquese**
https://improbableio.atlassian.net/browse/UNR-4845

This PR moves a bunch of logic that still existed in the sender into the relevant systems- actors, rpcs, well known entity, etc. 

Also:
- Load balance handler is now a singleton per worker owned by the netdriver
- fixes a bunch of dependency chains in order to do the above
- static component view is mostly unused as of this PR, but will remove in a separate PR
- load balance handler now uses the auth subview to keep track of the current state of auth intent

This PR is a halfway house towards moving everything to within systems. The next logical steps are [moving system commands out of sender and receiver](https://improbableio.atlassian.net/browse/UNR-4848), [move other system functionality out of sender](https://improbableio.atlassian.net/browse/UNR-4850) (which this PR does a lot of), and [removing the static component view](https://improbableio.atlassian.net/browse/UNR-4844), then iterate towards making these systems cleaner once [startup checkpointing](https://improbableio.atlassian.net/browse/UNR-4849) is unblocked.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Utils/EntityPool.h [hunk 1]
```diff
@@ -25,7 +23,7 @@ class FTimerManager;
 
 DECLARE_LOG_CATEGORY_EXTERN(LogSpatialEntityPool, Log, All)
 
-DECLARE_DYNAMIC_MULTICAST_DELEGATE(FEntityPoolReadyEvent);
+DECLARE_DELEGATE(FEntityPoolReadyEvent);
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/EntityPool.h:26]: Should we make this a `DECLARE_MULTICAST_DELEGATE`? The user can do AddXXX functions so there are multiple handlers to this event.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h [hunk 2]
```diff
@@ -18,6 +20,9 @@ class USpatialNetDriver;
 class SpatialActorChannel;
 class USpatialNetDriver;
 
+using FChannelsToUpdatePosition =
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:23]: The coding standard suggests against using `using` declarations at global scope: https://docs.unrealengine.com/en-US/ProductionPipelines/DevelopmentSetup/CodingStandard/#namespaces

> * Do not put using declarations in the global scope, even in a .cpp file (it will cause problems with our "unity" build system.)

Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:23]: Also, I wonder about the KeyFuncs passing `false` to `bInAllowDuplicateKeys` since we seem to always use it - can we specify it as a default argument? Isn't related to the immediate PR, of course.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/WellKnownEntitySystem.h [hunk 3]
```diff
@@ -28,10 +32,11 @@ class WellKnownEntitySystem
 
 	void InitializeVirtualWorkerTranslationManager();
 	void MaybeClaimSnapshotPartition();
+	void RetryServerWorkerEntityCreation(Worker_EntityId EntityId, int AttemptCounter);
 
 	const FSubView* SubView;
 
-	USpatialReceiver* Receiver;
+	USpatialNetDriver* NetDriver;
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/WellKnownEntitySystem.h:39]: Do we generally want to use `TWeakObjectPtr` to see if an object has been destroyed? I haven't seen many such usages in SpatialGDK code.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLoadBalancingHandler.h [hunk 4]
```diff
@@ -2,30 +2,49 @@
...
+	NotReady,
+	PendingKill,
+	NotInitialized,
+	Streaming,
+	NetDormant,
+	NoSpatialClassFlags,
+	DormantOnConnection
+};
+
 class FSpatialLoadBalancingHandler
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLoadBalancingHandler.h:31]: Shouldn't this class be inside `SpatialGDK` namespace? 🤔

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h [hunk 5]
```diff
@@ -31,5 +31,5 @@ class SPATIALGDK_API USpatialReplicationGraph : public UReplicationGraph
 	virtual void PostReplicateActors(UNetReplicationGraphConnection* ConnectionManager) override;
 	//~ End UReplicationGraph Interface
 
-	TUniquePtr<FSpatialLoadBalancingHandler> LoadBalancingHandler;
+	FSpatialLoadBalancingHandler* LoadBalancingHandler;
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h:34]: How about we use a TWeakPtr<FSpatialLoadBalancingHandler> here...

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 6]
```diff
@@ -203,6 +188,8 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	TUniquePtr<SpatialGDK::WellKnownEntitySystem> WellKnownEntitySystem;
 	TUniquePtr<SpatialGDK::ClientConnectionManager> ClientConnectionManager;
 
+	TUniquePtr<FSpatialLoadBalancingHandler> LoadBalancingHandler;
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: ...and a TSharedPtr<FSpatialLoadBalancingHandler> here?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp [hunk 7]
```diff
@@ -181,4 +197,104 @@ void WellKnownEntitySystem::MaybeClaimSnapshotPartition()
...
+	TArray<FWorkerComponentData> Components;
+	Components.Add(Position().CreateComponentData());
+	Components.Add(Metadata(FString::Format(TEXT("WorkerEntity:{0}"), { Connection->GetWorkerId() })).CreateComponentData());
+	Components.Add(ServerWorker(Connection->GetWorkerId(), false, Connection->GetWorkerSystemEntityId()).CreateServerWorkerData());
+
+	AuthorityDelegationMap DelegationMap;
+	DelegationMap.Add(SpatialConstants::GDK_KNOWN_ENTITY_AUTH_COMPONENT_SET_ID, EntityId);
+	Components.Add(AuthorityDelegation(DelegationMap).CreateComponentData());
+
+	check(NetDriver != nullptr);
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:219]: Is this a duplicate of a check above?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h (hunk 6 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: If we make this a shared pointer, is that just to enable using weak pointer in the rep graph or is there more benefit? Even so, it does make sense, will do this

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h (hunk 5 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h:34]: addressed below

File: SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp (hunk 7 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:219]: ha, yeah, didn't read the code I was copy pasting as thoroughly as I should have 😄

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h (hunk 6 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: It would enable actual shared ownership if we ever want - but so far it's just to enable TWeakPtr here https://github.com/spatialos/UnrealGDK/pull/2879#discussion_r572854668 so possible future crashes are less confusing :)

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: makes sense- although personally I'm not sure shared ownership is a benefit here. Will add a comment explaining why this is a shared pointer.

Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: It seems to be an established pattern to use UniquePtr's for storing an object on the NetDriver, and raw pointers to keep a reference to the actual object; is there a reason behind this approach?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/EntityPool.h (hunk 1 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/EntityPool.h:26]: I made this a normal delegate because there was only one user AFAICT, and that seemed more explicit. Happy to change though

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h (hunk 6 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: I mean, I haven't seen a place where shared ownership would be required in our systems so far, but keeping a raw pointer to something on a completely different object, the lifetime of which is controlled by Unreal's GC, is scary nonetheless. 

I'm also pondering whether we can skip caching pointers to UniquePtr-owned systems on other objects and instead picked them up from the NetDriver every time we need them, as this would allow us to see that the NetDriver has expired.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/WellKnownEntitySystem.h (hunk 3 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/WellKnownEntitySystem.h:39]: Depends on the use case. For this one, the well known entity system has a well-defined lifecycle wrt the objects these members are pointing at, so I think a raw pointer here makes sense. 

To answer the actual question, I'm not sure there are any general rules we follow in the GDK for this kind of thing, unfortunately.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 8]
```diff
@@ -256,7 +256,7 @@ void USpatialActorChannel::RetireEntityIfAuthoritative()
 			}
 			else
 			{
-				Sender->RetireEntity(EntityId, Actor->IsNetStartupActor());
+				NetDriver->ActorSystem->RetireEntity(EntityId, Actor->IsNetStartupActor());
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:259]: Changes such as these seem rather unfortunate to me as ActorSystem (in my impression at least) has a single responsibility of mapping Spatial's entities conveniently to UE's actors, and isn't the one responsible for retiring entities, or updating interest (below in this file). Do we have a task to separate this, too?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 9]
```diff
@@ -1281,9 +1280,20 @@ void USpatialActorChannel::UpdateSpatialPosition()
...
 void USpatialActorChannel::SendPositionUpdate(AActor* InActor, Worker_EntityId InEntityId, const FVector& NewPosition)
 {
-	if (InEntityId != SpatialConstants::INVALID_ENTITY_ID && NetDriver->HasServerAuthority(InEntityId))
+	if (InEntityId != SpatialConstants::INVALID_ENTITY_ID)
 	{
-		Sender->SendPositionUpdate(InEntityId, NewPosition);
+#if !UE_BUILD_SHIPPING
+		if (!NetDriver->Connection->GetView()[EntityId].Authority.Contains(SpatialConstants::SERVER_AUTH_COMPONENT_SET_ID))
+		{
+			UE_LOG(LogSpatialSender, Verbose,
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:1288]: Log category cleanup?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h (hunk 2 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:23]: great catch- moved within SpatialGDK

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h (hunk 6 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h:191]: I agree that keeping raw pointers around on objects which is controlled by Unreal's GC is scary. That's why generally we are moving away from UObjects etc. so that we have a better understanding of the lifecycles of our classes. My mental model is lifecycles of most GDK classes that aren't the raw connection or Spatial View (maybe also the view even) should be tied to the SpatialNetDriver if possible, as GDK logic generally is related to the connection to a deployment. However, as you point out here, this makes sense as a weak pointer because the rep graph is a UClass. (At least, I think that's the reason it makes sense)

Skipping caching pointers- absolutely, great idea, some of the refactoring I've done has moved towards that approach. See the ActorSystem. Unfortunately we're still mid process, but the kind of setup you're describing sounds like the same end goal. 

Does this answer the question?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:1288]: Great catch, wow, for some reason I thought this wouldn't compile so that would catch these. removed spatial sender log category and changed all these

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 8 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:259]: Great question, and I agree not all of these fit the Actor system's current responsibilities. I don't think these are too far off that this is egregious to the point that we shouldn't do it- i.e. both retiring entities and updating Interest involve interacting with entities in order to map Actors to entities- but this could be better. I think making this cleaner falls under https://improbableio.atlassian.net/browse/UNR-4851 (especially if that will be split into more tasks as it looks like it will currently).

File: SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h (hunk 2 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/ActorSystem.h:23]: For the other comment, I think it makes sense, and it's very small, so added as a drive by.

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLoadBalancingHandler.h (hunk 4 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLoadBalancingHandler.h:31]: yes

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLoadBalancingHandler.h:30]: MAYBE the enum shouldn't have been, but "migration" is a uniquely GDK concept so left it for now

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h (hunk 6 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Yeah this is awesome, thank you!

File: SpatialGDK/Source/SpatialGDK/Public/Utils/EntityPool.h (hunk 1 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/EntityPool.h]: That's fair, no one is stopping us from changing back if there's ever a need. I'll leave this up to you.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/WellKnownEntitySystem.h (hunk 3 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/WellKnownEntitySystem.h:39]: Makes sense.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 8 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:259]: Agree 💯 , thanks for the clarification.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 10]
```diff
@@ -300,12 +348,6 @@ void ActorSystem::HandleActorAuthority(const Worker_EntityId EntityId, const Wor
 
 	if (NetDriver->IsServer())
 	{
-		// TODO UNR-955 - Remove this once batch reservation of EntityIds are in.
-		if (Authority == WORKER_AUTHORITY_AUTHORITATIVE)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:304]: This removed just due to the reference to the ticket above?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 11]
```diff
@@ -1712,12 +1777,349 @@ FString ActorSystem::GetObjectNameFromRepState(const FSpatialObjectRepState& Rep
...
 
-bool ActorSystem::EntityHasComponent(const Worker_EntityId EntityId, const Worker_ComponentId ComponentId) const
+void ActorSystem::SendCreateEntityRequest(USpatialActorChannel* Channel, uint32& OutBytesWritten)
+{
+	UE_LOG(LogActorSystem, Log, TEXT("Sending create entity request for %s with EntityId %lld, HasAuthority: %d"),
+		   *Channel->Actor->GetName(), Channel->GetEntityId(), Channel->Actor->HasAuthority());
+
+	const Worker_RequestId RequestId = CreateEntity(Channel, OutBytesWritten);
+
+	NetDriver->Receiver->AddPendingActorRequest(RequestId, Channel);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1787]: Shouldn't these pending requests be handled by the actor system now?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 10 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:304]: I think this was removed because I'm pretty sure the code path can't be hit. Will double check, maybe this was overzealous

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 12]
```diff
@@ -1712,12 +1777,349 @@ FString ActorSystem::GetObjectNameFromRepState(const FSpatialObjectRepState& Rep
...
+
+		NetDriver->Connection->SendDeleteEntityRequest(EntityId, RETRY_UNTIL_COMPLETE);
+	}
+}
+
+void ActorSystem::SendComponentUpdates(UObject* Object, const FClassInfo& Info, USpatialActorChannel* Channel,
+									   const FRepChangeState* RepChanges, const FHandoverChangeState* HandoverChanges,
+									   uint32& OutBytesWritten)
+{
+	// SCOPE_CYCLE_COUNTER(STAT_SpatialSenderSendComponentUpdates);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1942]: Please add the stats back in 🙏

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 13]
```diff
@@ -1712,12 +1777,349 @@ FString ActorSystem::GetObjectNameFromRepState(const FSpatialObjectRepState& Rep
...
+		UE_LOG(LogActorSystem, Warning,
+			   TEXT("Trying to send component update but don't have authority! Update will be queued and sent when authority gained. "
+					"entity: %lld"),
+			   EntityId);
+
+		// It may be the case that upon resolving a component, we do not have authority to send the update. In this case, we queue the
+		// update, to send upon receiving authority. Note: This will break in a multi-worker context, if we try to create an entity that
+		// we don't intend to have authority over. For this reason, this fix is only temporary.
+		TArray<FWorkerComponentUpdate>& UpdatesQueuedUntilAuthority = UpdatesQueuedUntilAuthorityMap.FindOrAdd(EntityId);
+		UpdatesQueuedUntilAuthority.Append(ComponentUpdates);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: Didn't you remove processing this above?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 14]
```diff
@@ -1712,12 +1777,349 @@ FString ActorSystem::GetObjectNameFromRepState(const FSpatialObjectRepState& Rep
...
+		return;
+	}
+
+	for (int i = 0; i < ComponentUpdates.Num(); i++)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1995]: nit: use range based for loop

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp [hunk 15]
```diff
@@ -41,18 +45,29 @@ void SpatialRPCService::ProcessChanges(const float NetDriverTime)
...
 	{
-		LastProcessingTime = NetDriverTime;
+		LastIncomingProcessingTime = NetDriverTime;
 		ProcessIncomingRPCs();
 	}
+
+	if (NetDriverTime - LastOutgoingProcessingTime > GetDefault<USpatialGDKSettings>()->QueuedOutgoingRPCRetryTime)
+	{
+		LastOutgoingProcessingTime = NetDriverTime;
+		ProcessOutgoingRPCs();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: Why was this added? We have the incoming one because they may have been queued, but I'm not aware of a reason for queueing outgoing rpcs.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp [hunk 16]
```diff
@@ -52,6 +57,17 @@ void WellKnownEntitySystem::Advance()
 	}
 }
 
+void WellKnownEntitySystem::SendClaimPartitionRequest(Worker_EntityId SystemWorkerEntityId, Worker_PartitionId PartitionId) const
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:60]: What was the reasoning behind putting this logic in the well known entity system? Seems more NetDriver level. For instance on startup flow a server will create it's entity, claim it, and then use that interest to "complete" the well known entity system.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp (hunk 15 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: Added is a strong word- this is moved, and only looks different because of how we now choose to process the outgoings. (i.e. uses delta time rather than a timermanager). This logic is the same as it was in the Sender. Are you saying we can remove it?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialNetDriverLoadBalancingHandler.cpp [hunk 17]
```diff
@@ -29,44 +29,44 @@ void FSpatialNetDriverLoadBalancingContext::UpdateWithAdditionalActors()
 	}
 }
 
-EActorMigrationResult FSpatialNetDriverLoadBalancingContext::IsActorReadyForMigration(AActor* Actor)
+SpatialGDK::EActorMigrationResult FSpatialNetDriverLoadBalancingContext::IsActorReadyForMigration(AActor* Actor)
 {
 	if (!Actor->HasAuthority())
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialNetDriverLoadBalancingHandler.cpp:34]: Possibly add `using SpatialGDK::EActorMigrationResult` to save on code?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp (hunk 16 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:60]: Reasoning: Partitions are counted as well known entities for this system (potentially they shouldn't be. The definition of a well known entity is too vague, which is my bad, needs revisiting). A goal is to move all uses of commands to the various subview-based systems (last thing until receiver and sender can be deleted) and so I thought partition commands would make sense to be done in the well known entity system, as the well known entity system will probably transform into the "startup checkpoint system" as of https://improbableio.atlassian.net/browse/UNR-4849.

Maybe it would be better to have an explicit partition system which handles this kind of thing and outputs a signal for the well known entity system to know when it's time to be "completed". Not sure.

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:60]: Wait are partitions even counted as well known entities, I think I'm totally wrong

which system is responsible for partitions?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 18]
```diff
@@ -628,12 +670,12 @@ void ActorSystem::HandleDeferredEntityDeletion(const DeferredRetire& Retire) con
 {
 	if (Retire.bNeedsTearOff)
 	{
-		NetDriver->Sender->SendActorTornOffUpdate(Retire.EntityId, Retire.ActorClassId);
+		NetDriver->ActorSystem->SendActorTornOffUpdate(Retire.EntityId, Retire.ActorClassId);
```
Comment (improbable-dmitrii) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:673]: Do you think we can do this since it's the same object?
```suggestion
		SendActorTornOffUpdate(Retire.EntityId, Retire.ActorClassId);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialNetDriverLoadBalancingHandler.cpp (hunk 17 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialNetDriverLoadBalancingHandler.cpp:34]: possibly a good idea - done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 18 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:673]: yep! And made the function const, thanks

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 11 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1787]: These are related to commands- so will be moved over to the Actor System as of https://improbableio.atlassian.net/browse/UNR-4848

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 12 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1942]: ah fine, done

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 14 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1995]: good plan

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 13 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: yes but now I'm not sure that was the best plan, need to think about this- having trouble thinking through if this is valid or not

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: with component sets we should only be trying to send updates if we have authority, there's no more waiting for authority due to dynamic components for example. 
But then the comment mentions sending updates to entities that don't exist- why would that happen? If the entity migrates away mid tick? Do we really want to queue up this update? When we check out the entity again the state of the entity may have changed because this can only happen if someone else gets authority, and our update may be invalid now, and we're not checking that here. It seems we should be trusting the logic which handles checking out new entities to deal with migrated entities that are coming back in the same way as entities that have migrated here for the first time.
And then in that case- what did this code ever do? What's the concrete bug that it fixes? Do you know, and will it still be a problem now?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp (hunk 15 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: Ahh we gotta queue rpcs while entity creation is in progress. I guess that means we have to stick this around. I'm not sure why this is on a timer though, I thought it was just flushed at the end of every frame.

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: we could change to be flushed every frame I guess- also doesn't this mean we could flush before the entity creation is complete right now, if this is only on a timer? Are we hoping creations come back before next flush?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 13 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: What about the time between a create entity request, and receiving the entity add? In that instance, we keep Unreal authority, but we don't have an entity we can make updates on, so they need to be queued.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp (hunk 16 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:60]: Partitions can be many entity types currently, player controllers, worker entities, generic partition entities. The only well known partition entity is the one stored in the snapshot.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:60]: Trying to think how this could be incorporated into a single system, and I'm not sure it's necessary. It's more that partitions are used across multiple systems, and the commands required to orchestrate them could come from multiple sources.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 13 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: why do we keep unreal authority? Can we make the entity creation/actor spawning synchronous?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: To simplify usability. This worker *will* have authority, and especially in a layered scenario where you never expect to lose authority, it's easier for the user to not have to deal with this. For them there's no distinction in operating on the actor before you create the entity, and before you check the entity out.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp (hunk 15 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: I believe we attempt to flush whenever we try to send an rpc currently, which might not be every frame, but most likely is.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: I think we changed the commands one to flush every frame and can probably do the same here. Currently we flush every frame by coincidence which isn't great. As when you don't flush every tick it doesn't do what you expect.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp:57]: Although the rpc refactor etc will also do this.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 13 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1991]: It's the reason we added OnReady as as solution that's easy to ignore. We can't make this synchronous so if you do ignore OnReady then things might go wrong.

So actors keep authority but you aren't supposed to do anything with that actor that might interact with another entity until the entity is created.

In the case where you are just using offloading and you don't care about worker recovery or snapshots then you can ignore OnReady safely (except in the case of sending rpcs to startup actors). So it specifically improves usability there, while increasingly flexibility but at the cost of a pitfall elsewhere.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp (hunk 16 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/WellKnownEntitySystem.cpp:60]: I agree- I think partitions should likely be handled by whatever system cares about that partition. This will be gone as of strategy worker anyway though, so I am leaning towards this not being too important for the minute

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 10 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:304]: going to add this back for the reasons discussed in the other thread (actions taken between actor creation and entity creation). Not sure if the mentioned ticket is relevant anymore https://improbableio.atlassian.net/browse/UNR-955, so will drop reference to that and add a comment about what this is now doing

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp (hunk 13 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2013]: added back the pre-auth buffering

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:2013]: Chatted about this offline - I was incorrect about us processing updates whilst the entity creation is in progress (we check against that inside IsReadyForReplication) - so seems safe to remove this queuing logic all together, and replace it with an error message.

**Comment by jacquese** [comment]
PR ended up with too many conflicts for me. This has been superseded by other PRs: https://github.com/spatialos/UnrealGDK/pull/2929, https://github.com/spatialos/UnrealGDK/pull/2943, and more to come.
